lib/mayaUsd/ufe/UsdStageMap.cpp

```diff
@@ -138,7 +142,12 @@ MObject UsdStageMap::proxyShape(const Ufe::Path& path)
 
     auto iter = fPathToObject.find(singleSegmentPath);
     if (iter != std::end(fPathToObject)) {
-        return iter->second.object();
+        // If the proxyShape has been re-parented in Maya & the stage lookup
```

Comment:
This fixes the issue I am hitting in internal code review 253390 where I can get a Ufe::SceneItem for a prim which had its proxyShape node re-parented using a Ufe::Path to the old proxy shape.

Does this lose the performance benefit of the cache? I'm not sure what exactly the cache was meant to speed up, but now the code is getting the DAG path of the object and converting that into a Ufe::Path every time we hit the cache.

Reply:
The comment on line 153 is correct. UsdStageMap is not listening for re-parent notifications for when the proxy shape is re-parented, so when a proxy shape is re-parented the cache remains valid and will respond to queries.

Following a re-parent operation the code will detect the invalid cache state when a query is made for the new proxy shape path which is not in the cache. That triggers the code at line 160 to execute a rebuild the cache.

The problem is if no queries have come in for the new proxy shape path then the old cache is still intact and handling queries. If a query comes in for the _old_ proxy shape path then there will be a cache hit & a proxy shape returned for a DAG path which doesn't exist! The cached MObject is valid, but it's path is not the path that was searched for.

The new code handles this case by verifying that the path of the MObject stored in the cache matches the cache key used to look up the MObject. These two paths will be different when the MObject has been re-parented & the cache hasn't been rebuilt yet.

Reply:
I think the assumption is that this cache was never meant to be called with the "old" name (after a reparent or rename). Is the performance impact of this change? I wonder if this UsdStageMap should be listening for Maya rename/reparent and if on any shape we've cache'd we invalidate cache.

Reply:
The performance impact is every time we hit the change we do two operations whose complexity is on the order of the length of a path to a proxy shape. So the cost is not very high, but it is not zero.

I think UsdStageMap originally had a callback based design but it was changed to the passive error detection approach for correctness. In the comments in the file it talks about other things such as outliner listening for the callbacks & trying to use UsdStageMap and there being a "race" between which callback is invoked first (because there is no ordering).

Even the current design is not 100% correct. If there was a series of re-parenting & renaming operations that swap the paths of two proxy shapes without allowing any queries in between. In this case the cache would never get rebuilt because there is always a cache hit, and the cache would return the wrong proxy shape to the user!

For example if we have two proxy shapes `|stage1|stageShape1` which has stage A and `|stage2|stageShape2` that has stage B. Then if I execute some python like:

```
import maya.cmds as cmds
cmds.rename('|stage1|stageShape1', '|stage1|intermediate')
cmds.rename('|stage2|stageShape2', '|stage2|stageShape1')
cmds.reparent('stage2|stageShape1', |stage1', relative=True)
cmds.reparent('stage1|intermediate', |stage2', relative=True)
cmds.rename('|stage2|intermediate', |stage2|stageShape2)
```

Then the cache will have `|stage1|stageShape1` return stage B and `|stage2|stageShape2` return stage A!

So really I think we need the code I added to handle the callback race case as best as we can, but we also need to listen to the paths again so that we can prevent errors from occurring.

Reply:
@williamkrick I'm on vacation for 1.5 weeks. I would suggest putting Krystian as reviewer in my place.